server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java

```diff
@@ -287,24 +288,40 @@ public boolean assignVMToBackupOffering(Long vmId, Long offeringId) {
-                    "Backup-" + vm.getHostName() + "-" + vm.getUuid(), vm.getBackupOfferingId(), null, null,
-                    Backup.class.getSimpleName(), vm.getUuid());
-        } else {
-            throw new CloudRuntimeException("Failed to update VM assignment to the backup offering in the DB, please try again.");
-        }
+        return transactionAssignVMToBackupOffering(vm, offering, backupProvider) != null;
+    }
 
-        try {
-            if (backupProvider.assignVMToBackupOffering(vm, offering)) {
-                return vmInstanceDao.update(vm.getId(), vm);
+    private VMInstanceVO transactionAssignVMToBackupOffering(VMInstanceVO vm, BackupOfferingVO offering, BackupProvider backupProvider) {
+        return Transaction.execute(TransactionLegacy.CLOUD_DB, new TransactionCallback<VMInstanceVO>() {
+            @Override
+            public VMInstanceVO doInTransaction(final TransactionStatus status) {
+                try {
+                    long vmId = vm.getId();
+                    vm.setBackupOfferingId(offering.getId());
+                    vm.setBackupVolumes(createVolumeInfoFromVolumes(volumeDao.findByInstance(vmId)));
+
+                    if (vmInstanceDao.update(vmId, vm)) {
+                        if (backupProvider.assignVMToBackupOffering(vm, offering)) {
+                            vmInstanceDao.update(vmId, vm);
+                            UsageEventUtils.publishUsageEvent(EventTypes.EVENT_VM_BACKUP_OFFERING_ASSIGN, vm.getAccountId(), vm.getDataCenterId(), vmId,
+                                    "Backup-" + vm.getHostName() + "-" + vm.getUuid(), vm.getBackupOfferingId(), null, null,
+                                    Backup.class.getSimpleName(), vm.getUuid());
+                            LOG.debug(String.format("VM [%s] successfully added to Backup Offering [%s].", ReflectionToStringBuilderUtils.reflectOnlySelectedFields(vm, "uuid", "instanceName", "backupOfferingId", "backupVolumes"),
+                                    ReflectionToStringBuilderUtils.reflectOnlySelectedFields(offering, "uuid", "name", "externalId", "provider")));
+                            return vm;
+                        }
+                    } else {
+                        throw new CloudRuntimeException("Failed to update VM assignment to the backup offering in the DB, please try again.");
+                    }
+                } catch (Exception e) {
+                    LOG.error(String.format("Failed to assign VM [%s] to the Backup Offering [%s], using provider [name: %s, class: %s], due to: [%s].",
+                            ReflectionToStringBuilderUtils.reflectOnlySelectedFields(vm, "uuid", "instanceName", "backupOfferingId", "backupVolumes"),
+                            ReflectionToStringBuilderUtils.reflectOnlySelectedFields(offering, "uuid", "name", "externalId", "provider"),
+                            backupProvider.getName(), backupProvider.getClass().getSimpleName(), e.getMessage()), e);
```

Comment:
I would not want a stack trace in an error message.
How about creating the message twice, once as error and once with stacktrace as debug?

Reply:
note btw, that this pokemon catch catches the exception thrown in the try-block. not a very nice pattern.

Reply:
I did not understand. For me it makes more sense for the stacktrace to appear in an error log than in a debug, in addition, this message will only be visible to operators, and the only error message the user will see is "Failed to assign the VM to the backup offering, please try removing the assignment and try again." or "Failed to update VM assignment to the backup offering in the DB, please try again.".

Reply:
@SadiJr I would like to see only the exception message in the error log. If the problem persisyst an I need to invetsigate I can up the level to debug, to look at stacktraces. The stacktraces are never usefull for operational errors, only for bug hunting. The average operator will not have a clue what to do with them.

Reply:
@DaanHoogland Now I understand your suggestion. And, I have to admit that you are completely right about stacktraces not being useful to operators. I applied your suggestions, thanks.